src/Voron/Impl/PagerState.cs

```diff
@@ -141,5 +182,93 @@ public void DebugVerify(long size)
         }
 
         public AbstractPager CurrentPager => _pager;
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetSegmentState(long segment)
+        {
+            if (segment < 0)
+                return AlreadyPrefetch;
+
+            byte value = this._prefetchTable[segment / 2];
+            if (segment % 2 == 0)
+            {
+                // The actual value is in the high byte.
+                value = (byte)(value >> EvenPrefetchMaskShift);
+            }
+            else
+            {
+                value = (byte)(value & OddPrefetchCountMask);
+            }
+
+            return value;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetSegmentState(long segment, int state)
+        {
+            byte value = this._prefetchTable[segment / 2];
+            if (segment % 2 == 0)
+            {
+                // The actual value is in the high byte.
+                value = (byte)((value & OddPrefetchCountMask) | (state << EvenPrefetchMaskShift));
+            }
+            else
+            {
+                value = (byte)((value & EvenPrefetchCountMask) | state);
+            }
+
+            this._prefetchTable[segment / 2] = value;
+        }
+
+        public bool ShouldPrefetchSegment(long pageNumber, out void* virtualAddress, out long sizeInBytes)
+        {
+            long segmentNumber = (pageNumber * Constants.Storage.PageSize) >> this._segmentShift;
+
+            int segmentState = GetSegmentState(segmentNumber);
+            if (segmentState < AlreadyPrefetch)
+            {
+                // We update the current segment counter
+                segmentState++;
+
+                int previousSegmentState = GetSegmentState(segmentNumber - 1);
```

Comment:
Need to also go over the  next segment as well, to make sure that we are not going backward.

Reply:
We are not, forgot about that one... in a sec